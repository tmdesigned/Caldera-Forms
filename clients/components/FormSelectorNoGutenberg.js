import React from 'react';
import PropTypes from 'prop-types';

import { FormGroup,HelpBlock } from 'react-bootstrap';

/**
 * Form selector with no dependency on Gutenberg
 *
 * @since 1.7.0
 *
 * @param props
 * @returns {*}
 * @constructor
 */
export const FormSelectorNoGutenberg = (props) => {
    const idAttr = props.idAttr ? props.idAttr :'caldera-forms-form-chooser';
    return (
        <FormGroup controlId={idAttr}>
            <label
                htmlFor={idAttr}
            >
                Choose Form
            </label>
            <select
                id={idAttr}
                value={props.selected || ''}
                className="select"
                onChange={(e) => {
                    props.onChange(e.target.value)
                }}
            >
                <option value=''></option>
                {Object.keys(props.forms).map( (formsIndex) => {
                    const form = props.forms[formsIndex];
                    return (
                        <option
                            key={form.ID}
                            value={form.ID}
                        >
                            {form.name}
                        </option>
                    )
                })}
            </select>
            <HelpBlock>Choose a form to edit privacy settings for.</HelpBlock>
        </FormGroup>
    )
};

FormSelectorNoGutenberg.propTypes = {
    forms: PropTypes.array.isRequired,
    onChange: PropTypes.func.isRequired,
    selected: PropTypes.string,
    idAttr: PropTypes.string
};