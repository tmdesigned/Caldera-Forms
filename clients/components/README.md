# Shared Components
This is a library of React components that can safely be shared by any React-driven UI, Gutenberg or not and are not specific to a client.


## Components
* `FormSelectorNoGutenberg` A react-bootstrap dependent selector for all saved forms.
* `CalderaHeader` A component to create the header markup of a page.
- Child props can be passed. They will be outputted inside of a `ul`. You must supply `li`.
* `PageBody` Wraps the content of an admin page in a consistent wrapper.
* `StatusIndicator` Conveys succesful green messages or red messages of failure.

### `StatusIndicator`
* Conveys successful green messages or red messages of failure.
* Relies on CSS in admin.css for style, which makes it look the same as existing similar components.
* Designed to work the same way as status indicator used in Pro UI, which is VueJS.

#### Examples
Success message:
```jsx harmony
<StatusIndicator 
    message={'Everything Is Sivan!'}
    show={true} 
    success={true}
/>
```

Error/ warning:

```jsx harmony
<StatusIndicator
    message={'Error!'}
    show={true}
    success={false}
/>

```