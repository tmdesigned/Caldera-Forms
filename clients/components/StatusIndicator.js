import React from  'react';
import PropTypes from 'prop-types';
export  const StatusIndicator = (props) =>{
    if( ! props.show){
        return null;
    }
    let className = 'cf-alert';
    if( props.success){
        className = `${className} cf-alert-success`;
    }else{
        className = `${className} cf-alert-error`;
    }

    return(
        <div className={"cf-alert-wrap"}>
            <p
                className={className}
            >
                {props.message}
            </p>
        </div>
    )



};
StatusIndicator.propTypes = {
    message: PropTypes.string,
    success: PropTypes.bool,
    show: PropTypes.bool,
};