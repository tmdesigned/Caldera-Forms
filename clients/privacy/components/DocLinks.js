import React from 'react';
import { Twemoji } from 'react-emoji-render';
import {ListGroup, ListGroupItem } from 'react-bootstrap'
import PropTypes from 'prop-types';

/**
 * Shows the doc links in the Privacy settings page
 *
 * @since 1.7.2
 *
 * @param props
 * @return {*}
 */
export const DocLinks = (props) => {
    return (
        <React.Fragment>
            <ListGroup
                style={{padding:'12px'}}
            >
                <ListGroupItem>
                    <Twemoji text=":eyes:" />
                    <a href="https://calderaforms.com/doc/setting-caldera-forms-gdpr-data-requests/?utm_source=wp-admin&utm_campaign=privacy-settings">
                        Setting Up Caldera Forms For GDPR Requests
                    </a>
                </ListGroupItem>
                <ListGroupItem>
                    <Twemoji text=":eyes:" />
                    <a href="https://calderaforms.com/gdpr/?utm_source=wp-admin&utm_campaign=privacy-settings">
                       All Caldera Forms GDPR Compliance Tools
                    </a>
                </ListGroupItem>
				<ListGroupItem>
                    {props.children}
                </ListGroupItem>
            </ListGroup>
        </React.Fragment>
    )
};

DocLinks.propTypes = {
	extraLink: PropTypes.string,
	extraLinkText: PropTypes.string
};