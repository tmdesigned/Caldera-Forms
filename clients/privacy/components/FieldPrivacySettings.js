import React from 'react';
import PropTypes from 'prop-types';
import {FormGroup, FormControl, ControlLabel, HelpBlock, Panel, PanelGroup, Checkbox} from 'react-bootstrap';
import {IsEmailIdentifyingField} from "./IsEmailIdentifyingField";
import {IsPiiField} from "./IsPiiField";

/**
 * Privacy Settings for one field of a form
 *
 * @since 1.7.0
 *
 * @param props
 * @returns {*}
 * @constructor
 */
export const FieldPrivacySettings = (props) => {
    return (
        <section
            style={{
                backgroundColor: '#fff',
                margin: '1em',
            }}
            className={'row'}
        >
            <h4
                style={{
                    margin: 0,
                    padding: '.7em 1em',
                    borderBottom: '1px solid #eee'
                }}
            >
                {props.field.name}
            </h4>
            <div
                className={'col-xs-6'}
                style={{
                    padding: '12px'
                }}
            >
                <IsPiiField
                    field={props.field}
                    privacySettings={props.privacySettings}
                    onCheck={props.onCheckIsPii}
                />
            </div>

            <div
                className={'col-xs-6'}
                style={{
                    padding: '12px'
                }}
            >
                <IsEmailIdentifyingField
                    className={'col-xs-6'}
                    field={props.field}
                    privacySettings={props.privacySettings}
                    onCheck={props.onCheckIsEmail}
                />
            </div>


        </section>
    );
};

FieldPrivacySettings.propTypes = {
    field: PropTypes.object.isRequired,
    formId: PropTypes.string.isRequired,
    privacySettings: PropTypes.object.isRequired,
    onCheckIsEmail: PropTypes.func.isRequired,
    onCheckIsPii: PropTypes.func.isRequired
};