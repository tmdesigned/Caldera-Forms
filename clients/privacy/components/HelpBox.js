import React from 'react';
import PropTypes from 'prop-types';
import {Twemoji} from 'react-emoji-render';
import {DocLinks} from "./DocLinks";
import {ListGroup, ListGroupItem } from 'react-bootstrap'
export const HelpBox = (props) => {
	return (
		<div
			style={{
				backgroundColor: '#fff'
			}}
			className={'col-xs-4'}
		>
			{props.saveButton &&
				<p
					style={{
						margin: 0,
						padding: '.7em 1em',
						borderBottom: '1px solid #eee'
					}}
				>
					{props.saveButton}
				</p>

			}
			<h3>
				<Twemoji text=":volcano: Terms" />
			</h3>
			<ListGroup
				style={{padding:'12px'}}
			>
				<ListGroupItem>
					<Twemoji text=":eyes:"/>Email Identifying Fields: The field(s) of your form that can be used to determine whose data an entry belongs to.
				</ListGroupItem>
				<ListGroupItem>
					<Twemoji text=":eyes:"/>Personally Identifying Fields (PII): The field(s) of your form that contain PII about the person identified in the email identifying field.
				</ListGroupItem>
			</ListGroup>

			<h3>
				<Twemoji text=":volcano: Documentation" />
			</h3>
			<DocLinks/>
		</div>
	);
};

HelpBox.propTypes = {
	saveButton: PropTypes.oneOfType([
		PropTypes.node,
		PropTypes.object
	]),
};