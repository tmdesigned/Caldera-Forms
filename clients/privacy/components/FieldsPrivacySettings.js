import React from 'react';
import PropTypes from "prop-types";
import {FieldPrivacySettings} from "./FieldPrivacySettings";

/**
 * All field-related privacy settings for a form
 *
 * @since 1.7.0
 *
 * @param props
 * @returns {*}
 * @constructor
 */
export const FieldsPrivacySettings = (props) => {
    return (
        <div>
            {Object.keys(props.fields).map( (fieldId ) => {
                const field = props.fields[fieldId];
                if ( field && field.hasOwnProperty( 'ID' ) ) {
                    const idAttr = `caldera-forms-privacy-field-settings-${field.ID}`;
                    return (
                        <FieldPrivacySettings
                            id={idAttr}
                            key={idAttr}
                            field={field}
                            formId={props.formId}
                            privacySettings={props.privacySettings}
                            onCheckIsEmail={props.onCheckIsEmail}
                            onCheckIsPii={props.onCheckIsPii}
                        />
                    );
                } else {
                    <p>Invalid Field</p>
                }

            })}
        </div>
    )
};

FieldsPrivacySettings.propTypes = {
    fields: PropTypes.object.isRequired,
    formId: PropTypes.string.isRequired,
    privacySettings: PropTypes.object.isRequired,
    onCheckIsEmail: PropTypes.func.isRequired,
    onCheckIsPii: PropTypes.func.isRequired
};