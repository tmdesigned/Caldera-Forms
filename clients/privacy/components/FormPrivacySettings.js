import React from 'react';
import PropTypes from 'prop-types';
import {FormGroup, ControlLabel, Checkbox, HelpBlock, Button } from 'react-bootstrap';
import {FieldsPrivacySettings} from "./FieldsPrivacySettings";
import remove from 'lodash.remove';
import {HelpBox} from "./HelpBox";

/**
 * All privacy settings for a form
 *
 * @since 1.7.0
 *
 * @param props
 * @returns {*}
 * @constructor
 */
export const FormPrivacySettings = (props) => {
    /**
     * Toggle exporter enabled
     *
     * @since 1.7.0
     */
    const toggleEnable = () => {
        props.onStateChange({
            ...props.privacySettings,
            privacyExporterEnabled: !props.privacySettings.privacyExporterEnabled,
        });
    };

    /**
     * Add a field to list if in list, remove if not in list
     *
     * @since 1.7.0
     * @param {Array} fields
     * @param {String} fieldId
     * @returns {*}
     */
    function updateFieldsList(fields, fieldId) {
        if (!fields.length || !fields.includes(fieldId)) {
            fields.push(fieldId)
        } else {
            if( 1 === fields.length ){
                fields = [];
            }else{
                fields = remove(fields, (ID) => {
                    return ID === fieldId;
                });
            }

        }
        return fields;
    }

    /**
     * Toggle a field as an email field to identify users by or not.
     *
     * @since 1.7.0
     *
     * @param {String} fieldId
     */
    const toggleIsEmail = (fieldId) => {
        let emailIdentifyingFields = updateFieldsList(props.privacySettings.emailIdentifyingFields, fieldId);
        props.onStateChange({
            ...props.privacySettings,
            emailIdentifyingFields,
        });
    };

    /**
     * Toggle a field as a PII field or not
     *
     * @since 1.7.0
     *
     * @param fieldId
     */
    const toggleIsPii = (fieldId) => {
        let piiFields = updateFieldsList(props.privacySettings.piiFields, fieldId);
        props.onStateChange({
            ...props.privacySettings,
            piiFields,
        });
    };


    return (
        <section
            className={'layout-grid'}
        >
            <h2>{props.form.name}</h2>

            <div
                className={'row'}
            >
                <div
                    className={'col-xs-8'}
                >
                    <FormGroup controlId={`caldera-forms-privacy-gdpr-enable-${props.form.ID}`}>
                        <ControlLabel>
                            Enable GDPR Exporter For Form
                        </ControlLabel>
                        <Checkbox
                            onChange={toggleEnable}
                            checked={props.privacySettings.privacyExporterEnabled}
                        >
                            Enable
                        </Checkbox>
                        <HelpBlock>If checked data for this form will be added to GDPR personal data requests, and deletes.</HelpBlock>
                    </FormGroup>

                    {props.privacySettings.privacyExporterEnabled &&
                        <section>
                            <h3>Field Settings</h3>
                            <FieldsPrivacySettings
                                fields={props.form.fields}
                                formId={props.form.ID}
                                privacySettings={props.privacySettings}
                                onCheckIsEmail={toggleIsEmail}
                                onCheckIsPii={toggleIsPii}
                            />
                        </section>
                    }
                </div>
                <HelpBox
                    saveButton={
						<button
							type="submit"
							onClick={() => {
								props.onSave(props.privacySettings, props.form.ID)
							}}
							className={'primary button button-primary'}
						>
							Save
						</button>
                    }
                />
            </div>




        </section>
    );


};

FormPrivacySettings.propTypes = {
    form: PropTypes.object.isRequired,
    onSave: PropTypes.func.isRequired,
    privacySettings: PropTypes.shape({
        emailIdentifyingFields: PropTypes.array,
        piiFields: PropTypes.array,
        privacyExporterEnabled: PropTypes.bool,
        fields: PropTypes.object,
    }).isRequired,
    onStateChange: PropTypes.func.isRequired,
};