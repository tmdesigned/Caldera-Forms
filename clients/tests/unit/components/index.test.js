import renderer from 'react-test-renderer';
import {FormSelectorNoGutenberg} from "../../../components/FormSelectorNoGutenberg";
import {CalderaHeader} from "../../../components/CalderaHeader";
import {PageBody} from "../../../components/PageBody";
import {StatusIndicator} from "../../../components/StatusIndicator";

const forms  = [
	{
		ID: 'CFreal',
		name: 'Become Real'
	},
	{
		ID: 'CFunreal',
		name: 'Become Unreal'
	},

];
describe( 'Form selector without Gutenberg dependency', () => {
	it( 'Has form selection', () => {
		const formSelector = renderer.create(
			<FormSelectorNoGutenberg
				forms={forms}
				onChange={() => {}}
				idAttr={'reality-selector'}
				selected={'CFunreal'}
			/>
		);
		expect( formSelector.toJSON()).toMatchSnapshot();
	})
});

describe( 'Caldera Header', () => {
	it( 'Works with no props', () => {
		const simpleHeader = renderer.create(
			<CalderaHeader/>
		);
		expect( simpleHeader.toJSON()).toMatchSnapshot();

	});

	it( 'Works with no name prop', () => {
		const headerWithName = renderer.create(
			<CalderaHeader
				name={'Settings'}
			/>
		);
		expect( headerWithName.toJSON()).toMatchSnapshot();

	});

	it( 'Can have menu a option', () => {
		const headerWithMenuItems = renderer.create(
			<CalderaHeader
				children={<li>Item</li>}
			/>
		);
		expect( headerWithMenuItems.toJSON()).toMatchSnapshot();
	});

	it( 'Can have menu options', () => {
		const headerWithMenuItems = renderer.create(
			<CalderaHeader
				children={[
					<li key={'1'}>Item 1</li>,
					<li key={'2'}>Item 2</li>
				]}
			/>
		);
		expect( headerWithMenuItems.toJSON()).toMatchSnapshot();
	});

});

describe( 'Page body', () => {
	it( 'Can have children', () => {
		const pageBody = renderer.create(
			<PageBody
				children={[
					<div key={'1'}>Item 1</div>,
					<div key={'2'}>Item 2</div>
				]}
			/>

		);
		expect( pageBody.toJSON()).toMatchSnapshot();

	});
});

describe( 'Status indicator', () => {
	it( 'Shows success', () => {
		const success = renderer.create(
			<StatusIndicator
				message={'It worked'}
				success={true}
				show={true}
			/>

		);
		expect( success.toJSON()).toMatchSnapshot();

	});

	it( 'Shows failure', () => {
		const failure = renderer.create(
			<StatusIndicator
				message={'It did not work'}
				success={false}
				show={true}
			/>

		);
		expect( failure.toJSON()).toMatchSnapshot();
	});

	it( 'Shows nothing', () => {
		const nothing = renderer.create(
			<StatusIndicator
				show={false}
			/>

		);
		expect( nothing.toJSON()).toMatchSnapshot();
	});


});
