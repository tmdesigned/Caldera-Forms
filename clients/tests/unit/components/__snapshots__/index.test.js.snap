// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Caldera Header Can have menu a option 1`] = `
<div
  className="caldera-editor-header"
>
  <ul
    className="caldera-editor-header-nav"
  >
    <li
      className="caldera-editor-logo"
    >
      <span
        className="caldera-forms-name"
      >
        Caldera Forms: 
        Caldera Forms
      </span>
    </li>
    <li>
      Item
    </li>
  </ul>
</div>
`;

exports[`Caldera Header Can have menu options 1`] = `
<div
  className="caldera-editor-header"
>
  <ul
    className="caldera-editor-header-nav"
  >
    <li
      className="caldera-editor-logo"
    >
      <span
        className="caldera-forms-name"
      >
        Caldera Forms: 
        Caldera Forms
      </span>
    </li>
    <li>
      Item 1
    </li>
    <li>
      Item 2
    </li>
  </ul>
</div>
`;

exports[`Caldera Header Works with no name prop 1`] = `
<div
  className="caldera-editor-header"
>
  <ul
    className="caldera-editor-header-nav"
  >
    <li
      className="caldera-editor-logo"
    >
      <span
        className="caldera-forms-name"
      >
        Caldera Forms: 
        Settings
      </span>
    </li>
  </ul>
</div>
`;

exports[`Caldera Header Works with no props 1`] = `
<div
  className="caldera-editor-header"
>
  <ul
    className="caldera-editor-header-nav"
  >
    <li
      className="caldera-editor-logo"
    >
      <span
        className="caldera-forms-name"
      >
        Caldera Forms: 
        Caldera Forms
      </span>
    </li>
  </ul>
</div>
`;

exports[`Form selector without Gutenberg dependency Has form selection 1`] = `
<div
  className="form-group"
>
  <label
    htmlFor="reality-selector"
  >
    Choose Form
  </label>
  <select
    className="select"
    id="reality-selector"
    onChange={[Function]}
    value="CFunreal"
  >
    <option
      value=""
    />
    <option
      value="CFreal"
    >
      Become Real
    </option>
    <option
      value="CFunreal"
    >
      Become Unreal
    </option>
  </select>
  <span
    className="help-block"
  >
    Choose a form to edit privacy settings for.
  </span>
</div>
`;

exports[`Page body Can have children 1`] = `
<div
  className="caldera-forms-admin-page-wrap"
  style={
    Object {
      "marginTop": "75px",
    }
  }
>
  <div>
    Item 1
  </div>
  <div>
    Item 2
  </div>
</div>
`;

exports[`Status indicator Shows failure 1`] = `
<div
  className="cf-alert-wrap"
>
  <p
    className="cf-alert cf-alert-error"
  >
    It did not work
  </p>
</div>
`;

exports[`Status indicator Shows nothing 1`] = `null`;

exports[`Status indicator Shows success 1`] = `
<div
  className="cf-alert-wrap"
>
  <p
    className="cf-alert cf-alert-success"
  >
    It worked
  </p>
</div>
`;
